### PR TITLE
[discovery.mdns] Devices may apply a grace period for removal from the Inbox

### LINKED ISSUE #1835: [MDNSDiscoveryService] potential bug deleting items from Inbox??
I have been developing a binding that implements a MDNSDiscoveryParticipant to discover the bridge. One user reports that every time the participant add such a bridge to the Inbox, the core then deletes it again. I think this could be a bug in the MDNSDiscoveryService class..

- the method `removeOlderResults()` method seems to be called, but it is not clear why.
- and also `timestampOfLastScan` seems not to be being initialized.

```
public class MDNSDiscoveryService extends AbstractDiscoveryService implements ServiceListener {
..
    protected synchronized void stopScan() {
        removeOlderResults(getTimestampOfLastScan());
        super.stopScan();
    }
```

### PR DESCRIPTION
**Pull request by andrewfg**
**BACKGROUND**

Some types of OpenHAB bindings have devices that can sometimes be a bit late in sending their mDNS renewal announcements, which means that such devices are repeatedly removed from, and (re)added to, the InBox. This leads to confusion in the UI, and repeated logger messages. This is an exact analog to the UPNP Discovery issue that was solved by https://github.com/openhab/openhab-core/pull/2144

**SOLUTION**

The solution is an exact analog to the solution in https://github.com/openhab/openhab-core/pull/2144
Resolves #1835

**DOCUMENTATION**

The binding developer guide will also be updated via this PR https://github.com/openhab/openhab-docs/pull/1708

Signed-off-by: Andrew Fiddian-Green software@whitebear.ch

### CODE REVIEW

**Comment by andrewfg** [comment]
Resolves https://github.com/openhab/openhab-core/issues/1835 in general.
And in combination https://github.com/openhab/openhab-addons/pull/11834 it resolves the specific Miele issue.

**Comment by jlaur** [comment]
@andrewfg - link #1835 to this PR?

**Comment by andrewfg** [comment]
^
Yes. See my post above yours.

**Comment by jlaur** [comment]
> ^ Yes. See my post above yours.

Ah, well, I meant reference it here to be able to directly cross-navigate and have issue auto-closed when PR is merged:

image

**Comment by andrewfg** [comment]
AFAIK the keyword “Resolves” (or “Fixes”, “Solves” etc.) followed by the issue number should indeed cause Git to auto close the issue.

**Comment by jlaur** [comment]
> AFAIK the keyword “Resolves” (or “Fixes”, “Solves” etc.) followed by the issue number should indeed cause Git to auto close the issue.

Yes, if you write it like "Fixes #1835" in the PR description, it will be linked correctly.

File: bundles/org.openhab.core.config.discovery.mdns/src/main/java/org/openhab/core/config/discovery/mdns/internal/MDNSDiscoveryService.java
Comment (jlaur) [inline bundles/org.openhab.core.config.discovery.mdns/src/main/java/org/openhab/core/config/discovery/mdns/internal/MDNSDiscoveryService.java:245]: Maybe this should be done unconditionally? If grace period in binding has changed to 0 since the task was scheduled, this might prevent the removal in case of any exception thrown.

File: bundles/org.openhab.core.config.discovery.mdns/src/main/java/org/openhab/core/config/discovery/mdns/internal/MDNSDiscoveryService.java
Comment (jlaur) [inline bundles/org.openhab.core.config.discovery.mdns/src/main/java/org/openhab/core/config/discovery/mdns/internal/MDNSDiscoveryService.java:243]: **Naming:** We have `ServiceEvent` and `ServiceInfo`. To avoid confusion, I would suggest to stick to this naming for both variables, i.e. `serviceEvent` (already named that) and `serviceInfo`, respectively.

File: bundles/org.openhab.core.config.discovery.mdns/src/main/java/org/openhab/core/config/discovery/mdns/internal/MDNSDiscoveryService.java
Comment (jlaur) [inline bundles/org.openhab.core.config.discovery.mdns/src/main/java/org/openhab/core/config/discovery/mdns/internal/MDNSDiscoveryService.java:257]: javadoc missing character:
```suggestion
    /**
```

File: bundles/org.openhab.core.config.discovery.mdns/src/main/java/org/openhab/core/config/discovery/mdns/internal/MDNSDiscoveryService.java
Comment (jlaur) [inline bundles/org.openhab.core.config.discovery.mdns/src/main/java/org/openhab/core/config/discovery/mdns/internal/MDNSDiscoveryService.java:69]: javadoc missing character:
```suggestion
    /**
```

File: bundles/org.openhab.core.config.discovery.mdns/src/main/java/org/openhab/core/config/discovery/mdns/internal/MDNSDiscoveryService.java
Comment (andrewfg) [reply bundles/org.openhab.core.config.discovery.mdns/src/main/java/org/openhab/core/config/discovery/mdns/internal/MDNSDiscoveryService.java:245]: I don't think so. Most bindings (except the Miele one) will not have overridden the getRemoval method, so the default method is called and returns 0. So on those most bindings, we definitely must not differ from prior behavior. I don't want to break any other bindings.

Comment (jlaur) [reply bundles/org.openhab.core.config.discovery.mdns/src/main/java/org/openhab/core/config/discovery/mdns/internal/MDNSDiscoveryService.java:245]: No harm done trying to remove it from the ConcurrentHashMap if it doesn't exist:
https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html#remove-java.lang.Object-

Of course it doesn't change anything as long as no bindings supports this method. The PR itself doesn't change anything in this case. But when bindings will start supporting this, we can run into this situation where grace period will be changed from value >0 to 0 while having a task running. So I think this problem can be easily avoided by removing the condition?

Comment (andrewfg) [reply bundles/org.openhab.core.config.discovery.mdns/src/main/java/org/openhab/core/config/discovery/mdns/internal/MDNSDiscoveryService.java:245]: Ok.

File: bundles/org.openhab.core.config.discovery.mdns/src/main/java/org/openhab/core/config/discovery/mdns/internal/MDNSDiscoveryService.java
Comment (andrewfg) [reply bundles/org.openhab.core.config.discovery.mdns/src/main/java/org/openhab/core/config/discovery/mdns/internal/MDNSDiscoveryService.java:257]: Ok.

File: bundles/org.openhab.core.config.discovery.mdns/src/main/java/org/openhab/core/config/discovery/mdns/internal/MDNSDiscoveryService.java
Comment (andrewfg) [reply bundles/org.openhab.core.config.discovery.mdns/src/main/java/org/openhab/core/config/discovery/mdns/internal/MDNSDiscoveryService.java:69]: Ok.

File: bundles/org.openhab.core.config.discovery.mdns/src/main/java/org/openhab/core/config/discovery/mdns/internal/MDNSDiscoveryService.java
Comment (andrewfg) [reply bundles/org.openhab.core.config.discovery.mdns/src/main/java/org/openhab/core/config/discovery/mdns/internal/MDNSDiscoveryService.java:243]: Ok.

File: bundles/org.openhab.core.config.discovery.mdns/src/main/java/org/openhab/core/config/discovery/mdns/internal/MDNSDiscoveryService.java
Comment (jlaur) [inline bundles/org.openhab.core.config.discovery.mdns/src/main/java/org/openhab/core/config/discovery/mdns/internal/MDNSDiscoveryService.java:219]: Sorry for late comment, didn't catch in first round: consider extracting this to a private method (e.g. `scheduleRemovalTask`), so you would have one method for scheduling task and another method for cancelling task. Even though it would only have one usage, it should still simplify this method by making it slimmer, thus improve readability.

Comment (andrewfg) [reply bundles/org.openhab.core.config.discovery.mdns/src/main/java/org/openhab/core/config/discovery/mdns/internal/MDNSDiscoveryService.java]: Ok.

**Comment by jlaur** [comment]
I have tested this PR in combination with openhab/openhab-addons#11834 and it works as expected - no more constant removal and readding of the Miele XGW 3000 gateway. From code perspective it also looks fine to me. Thanks again for this fix @andrewfg. Now will leave it to core maintainers to review.

**Review by kaikreuzer — APPROVED** [review]
lgtm, thanks!

**Comment by cweitkamp** [comment]
I am afraid we missed one place in the manual scan where we have to cancel the removal task.

https://github.com/openhab/openhab-core/blob/670c64b5c441d3a0408cd55927397c3a616251af/bundles/org.openhab.core.config.discovery.mdns/src/main/java/org/openhab/core/config/discovery/mdns/internal/MDNSDiscoveryService.java#L152-L172

I will provide a follow-up PR with a fix and some minor code refactoring.

**Comment by andrewfg** [comment]
Ok. Thanks.

**Comment by cweitkamp** [comment]
See https://github.com/openhab/openhab-core/pull/2647

**Pull request merged**
